overlord/snapstate/snapstate.go

```diff
@@ -782,10 +790,25 @@ func doInstall(st *state.State, snapst *SnapState, snapsup SnapSetup, compsups [
 	installSet := state.NewTaskSet(tasks...)
 	installSet.MarkEdge(prereq, BeginEdge)
 	installSet.MarkEdge(setupAliases, BeforeHooksEdge)
-	installSet.MarkEdge(setupSecurity, BeforeMaybeRebootEdge)
-	installSet.MarkEdge(linkSnap, MaybeRebootEdge)
-	installSet.MarkEdge(autoConnect, MaybeRebootWaitEdge)
-	installSet.MarkEdge(setAutoAliases, AfterMaybeRebootWaitEdge)
+
+	// restart related edges
+	installSet.MarkEdge(setupSecurity, BeforeMaybeRebootEdge) // this edge is just for tests
+	var rebootTask, waitRebootTask *state.Task
+	if setupKmodComponents == nil {
+		rebootTask = linkSnap
+		waitRebootTask = autoConnect
+	} else {
+		logger.Noticef("reboot will happen after set-up of kernel-modules")
+		rebootTask = setupKmodComponents
+		waitRebootTask = afterSetupKmodComps
+	}
+	installSet.MarkEdge(rebootTask, MaybeRebootEdge)
+	installSet.MarkEdge(waitRebootTask, MaybeRebootWaitEdge)
+	// Let tasks know if they have to do something about restarts
+	rebootTask.Set("prepare-restart", true)
+	waitRebootTask.Set("finish-restart", true)
+	installSet.MarkEdge(setAutoAliases, AfterMaybeRebootWaitEdge) // this edge is just for tests
+
 	if installHook != nil {
 		installSet.MarkEdge(installHook, HooksEdge)
 	}
```

Comment:
I'm a bit surprised that we don't need to add/change how we can restart.MarkTaskAsRestartBoundary ?

Reply:
I looked into that with some detail actually, but did not see a changed needed there as far as the edges are set for the right tasks. @Meulengracht might confirm or not this, I could have had an oversight.

Reply:
However this is not the case for Remodelling where `snapstate.arrangeSnapTaskSetsLinkageAndRestart`. Remodelling is using its own stitching of task-sets, so I don't know what the expectations surrounding this is there.

Reply:
+1 on that for a test to just ensure boundaries are set as expected